reference/type_traits/is_trivially_copyable.md

```diff
@@ -26,16 +26,41 @@ namespace std {
 ## 効果
 `is_trivially_copyable`は、`T`がトリビアルコピー可能な型であるならば[`true_type`](true_type.md)から派生し、そうでなければ[`false_type`](false_type.md)から派生する。
 
+ユーザーが関数を宣言したとき、`user-declared`であると言える。`= default`/`= delete`指定されたものも含む。
+
+`user-provided`とは、`user-declared`なものから`= default`/`= delete`指定されたものを除くものである。
+
 「トリビアルコピー可能な型」とは、「`std::memcpy()`可能な型である」と言い換えることもできる。これに分類される型は、以下の全ての条件を満たす必要がある：
 
+- 全てのコピー/ムーブ コンストラクタ/代入演算子はtrivialもしくは`= delete`指定されている
+- コピー/ムーブ コンストラクタ/代入演算子のうち少なくとも1つは`= delete`指定されていない
+- trivialで= delete指定されていないデストラクタを持つ
+
+これをもうすこし具体的に解釈する。
+
+まずそのクラス自体に対して次の条件を満たす必要がある。
+
+- デストラクターはvirtualではない
+- デストラクターは`= delete`指定されていない
+- デストラクタと全てのコピー/ムーブ コンストラクタ/代入演算子は`user-provided`ではない
+- virtual関数を持たず、virtual基本クラスも持たない
+- コピー/ムーブ コンストラクタ/代入演算子のうち少なくとも1つは`= delete`指定されていない
+
+さらに非staticメンバ変数にも上記の5条件が適応される。
+
+もし対象となるクラスが基底クラスを保つ場合(なにかを継承している場合)は、継承関係にあるクラスのうち基底クラス方向すべてのクラスが上記の5条件(非staticメンバ変数も)を満たす必要がある。
+
+### 注意
+
+トリビアルコピー可能の定義はかつては以下の全ての条件を満たすことを要求していた。
+
 - 非トリビアルなコピーコンストラクタを持っていないこと
 - 非トリビアルなムーブコンストラクタを持っていないこと
 - 非トリビアルなコピー代入演算子を持っていないこと
 - 非トリビアルなムーブ代入演算子を持っていないこと
 - トリビアルなデストラクタを持っていること
 
-トリビアル(trivial)な特殊関数(コンストラクタ、代入演算子、デストラクタ)とは、「ユーザー定義されない特殊関数」のことを意味する。
-
+しかし、デストラクタが削除されていてもトリビアルコピー可能となる問題があった。[CWG issue 1734](https://wg21.cmeerw.net/cwg/issue1734)によって定義が見直された。
```

Comment:
「この仕様修正はC++17の規格に適用されたが、この問題が発見された以降の標準ライブラリ実装では対応している可能性がある。」
とかも足しておくといいですかね。
このへんの書き方の統一は今後考える必要はありますが…。

Reply:
defectなのでどのISOで発行されたC++規格書に対してか？はあまり意味を持たないと 
[標準C++の欠陥解決は、過去のバージョンに遡って適用される - Faith and Brave - C++で遊ぼう](https://faithandbrave.hateblo.jp/entry/2017/08/10/120545)
にかかれていたので、それに従ってC++17の規格に適用された話も省いちゃったんですが、やっぱり省かないほうがいいんでしょうか？

Reply:
過去に遡るのも限度があるので、2013年に報告されたから少なくともC++11までは遡れないのですよね。なので、いつ判明したのかと、cpprefjpとしてどの規格に対応しているのかは書いておいたほうがほかの編集者がそれを見たときもわかりやすいのではないかと思います。

Reply:
たとえばC++17規格がでて最初のコンパイラ対応のリリースバージョンには直っているようなdefectだったら、バージョン注記の必要なくcpprefjp上の仕様を直してよいかと思います。
そういう素直な対応ができない場合は、実装のずれなどもあるので、できる限りの情報は必要かと思います。

Reply:
> 2013年に報告されたから少なくともC++11までは遡れないのですよね。

そうなんですか？知らなかった。

statusをみるとcd4となっていて、

> http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_active.html
> CD4: A DR/DRWP or Accepted/WP issue not resolved in C++14 but included in the Committee Draft advanced for balloting at the June, 2014 WG21 meeting.

という意味らしいんですが、じゃあC++14には遡及できるでしょうかね・・・？

遡及して適用している実装が確認できればいいんでしょうか・・・？

>https://groups.google.com/a/isocpp.org/d/msg/std-discussion/ilmqKwKPETs/u4Ty0PuoBwAJ
>
> The model used by cppreference is like that of implementers: we include in "C++xy" patches to defects in the C++xy standard as published. It is not useful for us to, say, claim that vector<int> v; is ill-formed before C++17, because the <int> would be parsed as a header-name preprocessing token according to the published standards prior to the resolution of CWG 2000.
>
> Since C++11 CWG has generally used a special designation for defect reports in its issue lists, but it doesn't say how far back the DRs reach. LWG doesn't use a special designation at all :( In such cases we examine the behavior of implementations to see if the resolution should be treated as a defect report and if so for which standard revisions.

Reply:
C++14の規格には適用できそうですね。
実装はさかのぼって調査が大変だと思うので「x年以降対応している可能性がある」と書いておけばひとまずよいかと思います。